File: DSA/LakshaKundra_590013237/Day_76_Question_2.py
class Node:
    def __init__(self, key):
        self.data = key
        self.left = None
        self.right = None

def print_ancestors(root, target):
    if not root:
        return False
    if root.data == target:
        return True
    if print_ancestors(root.left, target) or print_ancestors(root.right, target):
        print(root.data, end=" ")
        return True
    return False

# Example 1
root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)

print_ancestors(root, 4)  # Output: 2 1

print()

# Example 2
root = Node(10)
root.left = Node(20)
root.right = Node(30)
root.left.left = Node(40)
root.left.right = Node(50)

print_ancestors(root, 50)  # Output: 20 10

File: DSA/PriyankaSingh_590016827/day76question2.c
#include <stdio.h>
#include <stdlib.h>

// Definition for a binary tree node.
typedef struct TreeNode {
    int data;
    struct TreeNode* left;
    struct TreeNode* right;
} TreeNode;

// Function to create a new tree node
TreeNode* createNode(int data) {
    TreeNode* newNode = (TreeNode*)malloc(sizeof(TreeNode));
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Print Ancestors of a Node
int printAncestors(TreeNode* root, int target) {
    if (root == NULL) {
        return 0;
    }
    if (root->data == target) {
        return 1;
    }
    if (printAncestors(root->left, target) || printAncestors(root->right, target)) {
        printf("%d ", root->data);
        return 1;
    }
    return 0;
}

int main() {
    // Example 1
    TreeNode* root1 = createNode(1);
    root1->left = createNode(2);
    root1->right = createNode(3);
    root1->left->left = createNode(4);
    int target1 = 4;
    printf("Ancestors of %d: ", target1);
    printAncestors(root1, target1);
    printf("\n");

    return 0;
}
